rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,263 @@
+
+As each controller is responsible for part of the overall DNS record set for a given DNS name in a shared zones and potentially as values in a shared record and as neither the zone nor records are lockable, there will be scenarios where one controller overwrites some part of a zone/record with a stale view. This can happen if two or more controllers attempt to update the remote zone at the same time as each controller may have read the zone before another clusters has executed it's write allowing the zone to be updated without the knowledge of other actors that have also done a read. During this type of timing clash, effectively the last controller to write will have its changes stored. When it does validation, its changes will still be present and it will revert to a stable cycle. The other controllers involved, will see their changes missing in the remote zone during their validation check and so will pull and update the zone again (setting a new validation with a random jitter applied to increase the chances they don't clash again). Again the last controller to write will see its changes. So with this we can predict a worst case scenario of (num clashing controllers * (validation_loop + jitter)). However with adding in the jitter, it is likely that this will be a shorter period of time as multiple clusters will fall out of sync and should resolve their state within the min-max requeue interval rather than only ever one at a time. 
```

Comment:
Might be worth noting here that some DNS providers, depending on how they work, also try and mitigate against this kind of scenarios. Google for instance requires that during any update you first delete the existing record and you must do this by providing the exact contents of that record. In this case if it was out of date it would error and fall fall out of the reconcile loop and try again.

We still however need to do what is described here for cases where the provider does not do this.